feat(wm): add grid layout

javierportillo opened the pull request:
Adds a new layout based on the Grid layout on LeftWM.

![vlcsnap-2024-02-19-10h33m19s959](https://github.com/LGUG2Z/komorebi/assets/17260001/2537af5b-e8ee-4fde-a8ad-6c03677f2236)

Things I would Like to discuss that I'm not so sure about:
1. I needed to add the `op_direction` and `count` to the `Direction` trait to be able to calculate the grid neighbor, causing it to also be passed to the CustomLayout operations, but left unused.

2. Also, I had to add a couple of `#[allow(clippy::cast_possible_truncation)]` insted of actually solving the casting problems. I'm too much of a rust newbie to deal with that so I please tell me if I can do something about it, other than that I hope It is actually usable and shipable.

Thank you so much for Komorebi in general, I used to love using i3 on Arch (btw) but had to switch to Windows and I really missed a tiling WM, so Komorebi has been a blessing to me. Thanks!

javierportillo reviewed (commented):
Would this be a correct way of NoOping this commands?

javierportillo commented:
Hi, I just realized that I need to do something about the flip-layout command, it doesn't really work with this layout. Will be working on that...

LGUG2Z commented on komorebi/src/window_manager.rs:
```diff
@@ -1464,6 +1464,16 @@ impl WindowManager {
         let workspace = self.focused_workspace_mut()?;
+
+        // NoOp on Grid layout
+        match workspace.layout() {
```
You can use `matches!` here  since we are only ever comparing for one variant, and let's add some debug logs as well since this deviates from the established behaviour

```rust
if matches!(workspace.layout(), Layout::Default(DefaultLayout::Grid)) {
    tracing::debug!("ignoring promote command for grid layout");
    return Ok(())
}
```

LGUG2Z commented on komorebi/src/window_manager.rs:
```diff
@@ -1464,6 +1464,16 @@ impl WindowManager {
         tracing::info!("promoting container");
```
And let's bring this log down after the `matches!` block

LGUG2Z commented on komorebi/src/window_manager.rs:
```diff
@@ -1476,7 +1486,11 @@ impl WindowManager {
         let workspace = self.focused_workspace_mut()?;
         let target_idx = match workspace.layout() {
-            Layout::Default(_) => 0,
+            Layout::Default(layout) => match layout {
```
Here let's use the same `matches!` path as above to return early instead of engaging with this logic

LGUG2Z commented on komorebi-core/src/direction.rs:
```diff
@@ -19,10 +19,10 @@ pub trait Direction {
-    fn down_index(&self, idx: usize) -> usize;
-    fn left_index(&self, idx: usize) -> usize;
-    fn right_index(&self, idx: usize) -> usize;
+    fn up_index(&self, op_direction: OperationDirection, idx: usize, count: usize) -> usize;
```
Since this is ignored when implemented for `CustomLayout` let's make this an `Option<OperationDirection>` so that it's clear for different structs that implement `Direction` whether this is operated on or not

javierportillo commented:
Also, this would  close #256 👌

javierportillo commented:
Hey, I just discovered something else, when you are on any other layout and flip it, and then change (or cycle) to grid, it stays flipped and the direction movements are opposite of intended. I will try to implement the flip behaviour as the layout is not always symetrical, so I think it would be a good thing to have. If I can't implement it I will at try to come up with an alternative.

LGUG2Z commented:
> Hey, I just discovered something else, when you are on any other layout and flip it, and then change (or cycle) to grid, it stays flipped and the direction movements are opposite of intended.

Is this something specific to `grid` or does it happen to other layouts too? If it happens to other layouts (ie. it is pre-existing weird behaviour) I don't think you need to worry about it. I've had a look through the code and I think it looks good! Just waiting on your confirmation to merge this! 🚀

javierportillo commented:
I can't tell if it is intended behaviour to flip the layout and have it keep the flip state across changes (or cycles) to another layout, but it does happen with every layout. So I removed the early return from the flip_layout command and added support for flipping the grid, even if it is not that usefull to flip the grid but I think it better this way.

LGUG2Z commented:
Merged! 🎉

This is such a great PR and one of the best first-time contributions (and I believe this is also a first-time Rust contribution??) I've ever seen to an open source repository.

LGUG2Z commented:
Could you also open another little PR adding an ASCII diagram to match the others here? I think there might be one in LeftWM that we could reuse

https://github.com/LGUG2Z/komorebi/blob/master/docs/example-configurations.md?plain=1#L88

javierportillo commented:
Sure, I'm on it.

Thanks for the merge! 👌

LGUG2Z commented:
Actually don't worry about it I went ahead and stole the ASCII art from LeftWM as well 🤣 

https://github.com/LGUG2Z/komorebi/compare/master...docs/v0.1.22#diff-a2dea01d0b2b6da439e4515037a30766709ba7c1e9ca41a0e5b0071e23cb1b21R145
